agent/kafka_handler.go

```diff
@@ -17,6 +29,130 @@ type kafkaHandler struct {
+	if !validTopicChars.MatchString(name) {
+		return fmt.Errorf("'%s' contains one or more characters other than ASCII alphanumerics, '.', '_' and '-'", name)
+	}
+	return nil
+}
+
+func (k *kafkaHandler) HandleCreateTopicsRequest(hdr *kafkaprotocol.RequestHeader, req *kafkaprotocol.CreateTopicsRequest, completionFunc func(resp *kafkaprotocol.CreateTopicsResponse) error) error {
+	resp := &kafkaprotocol.CreateTopicsResponse{
+		ThrottleTimeMs: 0,
+		Topics:         make([]kafkaprotocol.CreateTopicsResponseCreatableTopicResult, len(req.Topics)),
+	}
+	for tidx, topic := range req.Topics {
+		retentionTime := k.agent.cfg.DefaultTopicRetentionTime
+		respConfigs := make([]kafkaprotocol.CreateTopicsResponseCreatableTopicConfigs, len(topic.Configs))
+		// Parse custom retention time from topic configs if provided
+		for cidx, config := range topic.Configs {
+			if common.SafeDerefStringPtr(config.Name) == "retention.ms" {
+				retentionMs, err := strconv.Atoi(common.SafeDerefStringPtr(config.Value))
+				if err == nil {
+					retentionTime = time.Duration(retentionMs) * time.Millisecond
+				}
+			}
+			respConfigs[cidx] = kafkaprotocol.CreateTopicsResponseCreatableTopicConfigs{
+				Name:  config.Name,
+				Value: config.Value,
+			}
+		}
+		var errMsg string
+		errCode := int16(0)
+		derefTopicName := common.SafeDerefStringPtr(topic.Name)
+		err := detectInvalidTopic(derefTopicName)
+		if err != nil {
+			errMsg = fmt.Sprintf("Invalid topic: %s (Reason: %s)\n", derefTopicName, err.Error())
+			errCode = int16(kafkaprotocol.ErrorCodeInvalidTopicException)
+		} else {
+			acl, err := k.agent.controlClientCache.GetClient()
+			if err != nil {
+				errMsg = err.Error()
+				errCode = kafkaprotocol.ErrorCodeCoordinatorNotAvailable
+			} else {
+				topicInfo := topicmeta.TopicInfo{
+					Name:           derefTopicName,
+					PartitionCount: int(topic.NumPartitions),
+					RetentionTime:  retentionTime,
+				}
+				err = acl.CreateTopic(topicInfo)
+				if err != nil {
+					errMsg = err.Error()
+					if extractErrorCode(err) == common.TopicAlreadyExists {
+						errCode = kafkaprotocol.ErrorCodeTopicAlreadyExists
+					} else {
+						errCode = kafkaprotocol.ErrorCodeInvalidTopicException
+					}
+				}
+			}
+		}
+		resp.Topics[tidx] = kafkaprotocol.CreateTopicsResponseCreatableTopicResult{
+			Name:          topic.Name,
+			ErrorCode:     errCode,
+			ErrorMessage:  &errMsg,
+			NumPartitions: topic.NumPartitions,
+			Configs:       respConfigs,
+		}
+	}
+	return completionFunc(resp)
+}
+
+func (k *kafkaHandler) HandleDeleteTopicsRequest(hdr *kafkaprotocol.RequestHeader, req *kafkaprotocol.DeleteTopicsRequest, completionFunc func(resp *kafkaprotocol.DeleteTopicsResponse) error) error {
+	resp := &kafkaprotocol.DeleteTopicsResponse{
+		ThrottleTimeMs: 0,
+		Responses:      make([]kafkaprotocol.DeleteTopicsResponseDeletableTopicResult, len(req.TopicNames)),
+	}
+	for tidx, topicName := range req.TopicNames {
+		var errMsg string
```

Comment:
should be *string otherwise we are setting err message to pointer to empty string even when there is no error

Reply:
I think this comment is outdated as the code has changed a bit. Please let me know if this is still needed.

Reply:
got it. Done.